### DISCUSSION TITLE
About pointer to class, what is the difference between CS8500 & CS0208?

### DISCUSSION THREAD

**Discussion by JJLovesLife** [dotnet/runtime#82664] [Q&A] [answered]
```
var obj = new object();
unsafe { Console.WriteLine(new IntPtr(&obj)); }
```
This code snippet will cause error CS0208 in SDK 6.0.406, `Cannot take the address of, get the size of, or declare a pointer to a managed type`, while causing warning CS8500 in SDK 7.0.200. `This takes the address of, gets the size of, or declares a pointer to a managed type`
Wanna know that is the difference between CS0208 and CS8500?

And in [this spec](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/language-specification/unsafe-code#223-pointer-types), it is said that:

> The intuitive rule for mixing of pointers and references is that referents of references (objects) are permitted to contain pointers, *but referents of pointers are not permitted to contain references.*

Does the scope extend in new version SDK?

I only see a relative breaking change about `System.TypedReference` in .NET 7, but I am not sure if this relate to my question🤔

**Comment by reflectronic** [comment, marked as answer]
Before C# 11, it was an error for a pointer to point to a managed type in all circumstances. In C# 11, most of the errors were turned into warnings. See [the associated proposal](https://learn.microsoft.com/en-us/dotnet/csharp/language-reference/proposals/csharp-11.0/low-level-struct-improvements#changes-in-unsafe-context):
> Pointer types are extended to allow managed types as referent type. Such pointer types are written as a managed type followed by a `*` token. They produce a warning.
>
> The address-of operator is relaxed to accept a variable with a managed type as its operand.
>
> The fixed statement is relaxed to accept *fixed_pointer_initializer* that is the address of a variable of managed type `T` or that is an expression of an *array_type* with elements of a managed type `T`.

The error is still used where managed types are strictly disallowed. For example,
```cs
public void M() 
{
    var z = stackalloc object[10];
}
``` 
will still emit an error CS0208, because `stackalloc` of a managed type is not supported by the runtime.

**Reply by Joe4evr** [reply]
Kinda hoping the language could/would relax the rules here for structs containing managed fields, but oh well.
